Examples.md

```diff
@@ -29,7 +31,20 @@ async with grpc.insecure_channel('localhost:50051') as channel:
     response = await stub.Hi(...)
 ```
 
-### Ideal
+### Option 3
+
+```Python
+ctx = grpc.Context() \
+    .with_timeout_secs(5.0) \
+    .append_outgoing_metadata('key', 'value')
+# or: ctx.with_deadline(time.time() + 5.0)
+
+async with grpc.insecure_channel('localhost:50051') as channel:
+    stub = helloworld_pb2_grpc.AsyncGreeterClient(channel)
+    response = await stub.Hi(ctx, helloworld_pb2.GreetRequest(...))
```

Comment:
In the vast majority of cases the proper context is the one received as an argument for server handler.
For other cases either `Context.BACKGROUND` or `Context.TODO` should be used.

Using background/todo context should be explicit as missing context object has a downside of not propagating deadline/cancellation/tracing info, so reviewer should have an ability to catch it.

Reply:
Assuming `Context.BACKGROUND` is a thread-local or something similar, why would that not be an appropriate choice for a default?

Reply:
Thread-local variables don't play nice with concurrency models, e.g. what is a behavior if one wants to create an intermediate context for creating custom tracing span?
If we allow users to specify/override thread-local context - we're facing with the same problems as regular thread locals (state is leaking across coroutines, not propagated to other threads, avoiding either of this problems requires both to carefully write the code and carefully review it).
If we don't allow users to do it - there is just no way to create custom intermediate spans.

Also having explicit `Context` has a nice benefit - it's clear if function does any i/o or not just from its definition based on the presence of context argument (though it may be not that important with asyncio due to async keyword).

Reply:
Internally we've used an approach of making `Context.TODO` to use thread-local state, which can only be set by the interceptor-like layer. It makes it explicit when user relies on thread-local state and by itself can be a warning to code reviewer that there may be some other code changes required.

Reply:
Agree on the thread-locals. But I don't think it's the same story for `contextvars`-style coroutine-local. In the vast majority of cases, I think these would do exactly what the user wants without having to lay down any extra plumbing. I don't see any danger of state leaking across coroutines. The only danger would be state *not* propagating if the user failed to manually propagate the context after a coroutine hop. However, I think that this is a minority of cases. Perhaps I should write up an example of my idea so we can compare things apples to apples.

Reply:
My main concerns of not doing the extra plumbing, so implicitly using the `ctx` available, are:

- Opening the door to future bugs since not only user data but also gRPC data is implicitly inherited and used.
- IMO Testing and reviewing process is harder, since the context is implicit.

Reply:
I agree with @gnossen here, `contextvars` provides sane semantics and extensibility, so I'm fine with using it instead of explicit `Context` object **and** instead of keyword arguments.
It's common python3 concept so will not alienate with other libraries/frameworks, as well as will be understandable to engineers familiar with gRPC libraries from other languages (c++, golang, java, etc).